### DISCUSSION TITLE
Issues with 'nonce' attribute

### DISCUSSION THREAD

**Discussion by balinomad** [livewire/livewire#3045] [7. Bugs (archived)]
v.2.4.2.

Nonce attribute is missing for `livewire.js`. Additionally, a space missing when 'nonce' attribute is present.
A possible solution is a the following change in `/src/LivewireManager.php` file (lines 270-271).

```
<script src="{$fullAssetPath}" data-turbo-eval="false" data-turbolinks-eval="false"></script>
<script data-turbo-eval="false" data-turbolinks-eval="false"{$nonce}>
```
could be
```
<script src="{$fullAssetPath}" data-turbo-eval="false" data-turbolinks-eval="false" {$nonce}></script>
<script data-turbo-eval="false" data-turbolinks-eval="false" {$nonce}>
```
Note the space prior {$nonce}.

**Comment by austenc** [comment]
Hey @navindex, could you elaborate on your need of the `nonce` property on that tag? Is it just the lack of a space that's causing you problems, or are you trying to satisfy some sort of CSP?

This has been mentioned before -- [see this comment from Caleb](https://github.com/livewire/livewire/pull/2485#issuecomment-784355989)

**Comment by balinomad** [comment]
Hi @austenc,
I'm using Livewire with [laravel-csp](https://github.com/spatie/laravel-csp) by calling it like
```
@livewireScripts(['nonce' => csp_nonce()])
```
from Blade templates. Their `csp_nonce()` helper generates the nonce and takes care of adding it to the header.
Having said that, it would be nice to have the same syntax for `@livewireStyles`, too.

I've tried different ways to create a strict CSP, however, the easiest way is when I set `script` to `strict-dynamic` and use nonce everywhere.

**Comment by Oha-you** [comment]
I'm having the same problem. Strict CSP is enabled on the website and Livewire inserts a raw `<script>` tag without a `nonce`: https://github.com/livewire/livewire/blob/master/src/LivewireManager.php#L275

An example of the current output is:
```html
<script src="/vendor/livewire/livewire.js?id=124876dasd2767" data-turbo-eval="false"
        data-turbolinks-eval="false"></script>
<script data-turbo-eval="false" data-turbolinks-eval="false"
        nonce="HkjAjSJHjas==">window.livewire = new Livewire({'nonce': 'HkjAjSJHjas=='});
    window.Livewire = window.livewire;
    window.livewire_app_url = '';
...
```
Now I have to either whitelist `/vendor/livewire/livewire.js` (that has a dynamic `id` at the end) or rewrite the whole `javaScriptAssets` function.

**Comment by Oha-you** [comment]
Update: my bad, in my case CSP was triggered by `<a>` tags having empty `href="javascript:void(0)"` attributes.

Having `'self' => true,` value under `'script-src'` block in `secure-headers.php` doesn't produce any errors and should be safe to use. But I agree that for consistency both tags should have a `nonce`.
